presto-geospatial/src/main/java/com/facebook/presto/geospatial/BingTileFunctions.java

```diff
@@ -0,0 +1,314 @@
+
+        BingTile leftUpperTile = latitudeLongitudeToTile(envelope.getYMax(), envelope.getXMin(), (int) zoomLevel);
+        BingTile rightLowerTile = latitudeLongitudeToTile(envelope.getYMin(), envelope.getXMax(), (int) zoomLevel);
+
+        // XY coordinates start at (0,0) in the left upper corner and increase left to right and top to bottom
+        long tileCnt = (rightLowerTile.getX() - leftUpperTile.getX() + 1) * (rightLowerTile.getY() - leftUpperTile.getY() + 1);
+
+        BlockBuilder blockBuilder = BIGINT.createBlockBuilder(new BlockBuilderStatus(), (int) tileCnt);
+        for (int x = leftUpperTile.getX(); x <= rightLowerTile.getX(); x++) {
+            for (int y = leftUpperTile.getY(); y <= rightLowerTile.getY(); y++) {
+                BingTile tile = new BingTile(x, y, (int) zoomLevel);
+
+                if (boundingBox || !tileToPolygon(tile).disjoint(geometry)) {
+                    BIGINT.writeLong(blockBuilder, tile.encode());
+                }
+            }
+        }
+
+        return blockBuilder.build();
+    }
+
+    private static Point tileXYToLatitudeLongitude(int tileX, int tileY, int zoomLevel)
+    {
+        double mapSize = mapSize(zoomLevel);
+        double x = (clip(tileX * TILE_PIXELS, 0, mapSize - 1) / mapSize) - 0.5;
+        double y = 0.5 - (clip(tileY * TILE_PIXELS, 0, mapSize - 1) / mapSize);
+
+        double latitude = 90 - 360 * Math.atan(Math.exp(-y * 2 * Math.PI)) / Math.PI;
+        double longitude = 360 * x;
+        return new Point(longitude, latitude);
+    }
+
+    private static BingTile latitudeLongitudeToTile(double latitude, double longitude, int zoomLevel)
+    {
+        double x = (longitude + 180) / 360;
+        double sinLatitude = Math.sin(latitude * Math.PI / 180);
+        double y = 0.5 - Math.log((1 + sinLatitude) / (1 - sinLatitude)) / (4 * Math.PI);
+
+        int mapSize = mapSize(zoomLevel);
+        int tileX = (int) clip(x * mapSize + 0.5, 0, mapSize - 1);
+        int tileY = (int) clip(y * mapSize + 0.5, 0, mapSize - 1);
+        return new BingTile(tileX / TILE_PIXELS, tileY / TILE_PIXELS, zoomLevel);
+    }
+
+    private static OGCGeometry tileToPolygon(BingTile tile)
+    {
+        Point upperLeftCorner = tileXYToLatitudeLongitude(tile.getX(), tile.getY(), tile.getZoomLevel());
+        Point lowerRightCorner = tileXYToLatitudeLongitude(tile.getX() + 1, tile.getY() + 1, tile.getZoomLevel());
+
+        Polyline boundary = new Polyline();
+        boundary.startPath(upperLeftCorner);
+        boundary.lineTo(lowerRightCorner.getX(), upperLeftCorner.getY());
+        boundary.lineTo(lowerRightCorner);
+        boundary.lineTo(upperLeftCorner.getX(), lowerRightCorner.getY());
+
+        final Polygon polygon = new Polygon();
+        polygon.add(boundary, false);
+
+        return OGCGeometry.createFromEsriGeometry(polygon, null);
+    }
+
+    /**
+     * @return true if the geometry and its bounding box match
```

Comment:
I am trying to understand what this method does. If `geometry` is a point then it always "matches", what does `match` mean in this context? Given this method does some "matching" the name of the method `isBoundingBox` is somewhat confusing.

Reply:
@nezihyigitbasi  , this functions checks whether the geometry is a point or a rectangle. In that case, the set of tiles that covers the geometry is the same as the set of tiles that covers its bounding box. Otherwise, the set of tiles that covers bounding box need to be refined to remove tiles which do not overlap with the geometry.

Reply:
@nezihyigitbasi , how about isPointOrRectangle for function name?

   /**
     * @return true if the geometry is a point or a rectangle
     */
    private static boolean isPointOrRectangle(OGCGeometry geometry, Envelope envelope)
    {